src/compiler/scala/tools/nsc/ast/parser/TreeBuilder.scala

```diff
@@ -98,26 +98,36 @@ abstract class TreeBuilder {
   def makeCaseDef(pat: Tree, guard: Tree, rhs: Tree): CaseDef =
     CaseDef(gen.patvarTransformer.transform(pat), guard, rhs)
 
-  /** Creates tree representing:
+  /** Create tree representing:
    *    { case x: Throwable =>
-   *        val catchFn = catchExpr
-   *        if (catchFn isDefinedAt x) catchFn(x) else throw x
+   *        scala.runtime.ScalaRunTime.catchIdentity(catchExpr).applyOrElse(x, scala.runtime.ScalaRunTime.thrower)
+   *        //The names `x` and `f` are fresh names.
+   *        //@inline def f[A](pf: PartialFunction[Throwable, A]) = pf
+   *        //f(catchExpr).applyOrElse(x, t => throw t)
    *    }
+   *
+   *  The `catchExpr` can be either a cases block, for which the cases
+   *  are used directly, or an arbitrary expr to be wrapped as shown.
+   *  The expr must conform to `PartialFunction[Throwable, ?]`.
    */
-  def makeCatchFromExpr(catchExpr: Tree): CaseDef = {
-    val binder   = freshTermName()
-    val pat      = Bind(binder, Typed(Ident(nme.WILDCARD), Ident(tpnme.Throwable)))
-    val catchDef = ValDef(Modifiers(ARTIFACT), freshTermName("catchExpr"), TypeTree(), catchExpr)
-    val catchFn  = Ident(catchDef.name)
-    val body     = atPos(catchExpr.pos.makeTransparent)(Block(
-      List(catchDef),
-      If(
-        Apply(Select(catchFn, nme.isDefinedAt), List(Ident(binder))),
-        Apply(Select(catchFn, nme.apply), List(Ident(binder))),
-        Throw(Ident(binder))
+  def makeCatchFromExpr(catchExpr: Tree): List[CaseDef] = catchExpr match {
+    case Match(EmptyTree, cases) => cases
+    case _ =>
+      val binder = freshTermName()
+      val pat    = Bind(binder, Typed(Ident(termNames.WILDCARD), Ident(typeNames.Throwable)))
+
+      //q"catchIdentity(pf).applyOrElse(t, (x: Throwable) => throw x)"
+      import gen.CODE.REF
+      val doit = Apply(
+        Select(
+          Apply(REF(currentRun.runDefinitions.catchIdentityMethod), catchExpr :: Nil),
```

Comment:
The parser needs to be threadsafe, so this would need to be eagerly initialized. We eagerly initialize all class/module symbols referenced in the parser in `forceSymbolsUsedByParser`. See 760df9843a910d6c3618e490c752eb03fb6924bd for the corresponding test to drive the change.

I would probably suggest selecting the method by name here to avoid finding the right spot to initialize the lazy val in `RunDefinition`s in the same manner as  `forceSymbolsUsedByParser`.